File: backend/app/models/expense.py
from datetime import date, datetime, timezone
from typing import Optional

from sqlalchemy import Boolean, Date, DateTime, Float, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.constants import ExpenseType
from app.database.base import Base


class ExpenseCategory(Base):
    __tablename__ = "expense_categories"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    type: Mapped[ExpenseType] = mapped_column(String(10), nullable=False)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)

    user = relationship("User", back_populates="expense_categories")
    expenses = relationship("Expense", back_populates="category")


class Expense(Base):
    __tablename__ = "expenses"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    category_id: Mapped[int] = mapped_column(ForeignKey("expense_categories.id"), nullable=False)
    amount: Mapped[float] = mapped_column(Float, nullable=False)
    description: Mapped[str] = mapped_column(String(255), default="")
    notes: Mapped[str] = mapped_column(String(1000), default="")
    expense_date: Mapped[date] = mapped_column(Date, nullable=False)
    is_recurring: Mapped[bool] = mapped_column(Boolean, default=False)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False)
    recurring_parent_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("expenses.id"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(timezone.utc)
    )

    user = relationship("User", back_populates="expenses")
    category = relationship("ExpenseCategory", back_populates="expenses")
    tags = relationship("Tag", secondary="expense_tags", back_populates="expenses")

File: frontend/src/types/expense.ts
import { Tag } from "@/types/tag";

export type ExpenseType = "need" | "want";

export interface ExpenseCategory {
  id: number;
  name: string;
  type: ExpenseType;
  is_default: boolean;
}

export interface Expense {
  id: number;
  category: ExpenseCategory;
  amount: number;
  description: string;
  notes: string;
  expense_date: string;
  is_recurring: boolean;
  is_archived: boolean;
  recurring_parent_id: number | null;
  tags: Tag[];
  goal_id: number | null;
}

export interface ExpenseCreatePayload {
  category_id: number;
  amount: number;
  description: string;
  notes?: string;
  expense_date: string;
  is_recurring?: boolean;
  tag_ids?: number[];
  goal_id?: number | null;
  clear_goal?: boolean;
}

export interface ExpenseImportItem {
  category_name: string;
  amount: number;
  description?: string;
  expense_date: string;
}

export interface ExpenseBulkImportResult {
  imported: number;
  skipped: number;
  errors: string[];
}
